app/src/org/commcare/tasks/templates/CommCareTask.java

```diff
@@ -34,7 +42,24 @@ public CommCareTask() {
     protected final Result doInBackground(Params... params) {
         //Never have to wrap the entirety of your task.
         try {
-            return doTaskBackground(params);
+            // show cancel button; don't dismiss dialog until setup's complete
+            enableCancelButton();
+            setup(params);
+
+            enableDismissDialogOnCancel();
+
+            if (isCancelled()) {
+                return null;
+            }
+            mainBackgroundWork(params);
+
+            if (isCancelled()) {
+                return null;
+            }
+
+            // hide cancel button during commit phase
+            disableCancelButton();
```

Comment:
This also needs to denote that "Nothing should process the 'cancel' signal anymore, even if the disable signal hasn't propogated out yet"